presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java

```diff
@@ -239,4 +246,88 @@ public boolean revokePrivileges(PrivilegeBag privilegeBag)
     {
         return client.revoke_privileges(privilegeBag);
     }
+
+    @Override
+    public void grantRole(String role, String granteeName, PrincipalType granteeType, String grantorName, PrincipalType grantorType, boolean grantOption)
+            throws TException
+    {
+        List<RolePrincipalGrant> grants = listRoleGrants(granteeName, granteeType);
+        for (RolePrincipalGrant grant : grants) {
+            if (grant.getRoleName().equals(role)) {
+                if (grant.isGrantOption() == grantOption) {
+                    return;
+                }
+                if (!grant.isGrantOption() && grantOption) {
+                    revokeRole(role, granteeName, granteeType, false);
+                    break;
+                }
+            }
+        }
+        createGrant(role, granteeName, granteeType, grantorName, grantorType, grantOption);
+    }
+
+    private void createGrant(String role, String granteeName, PrincipalType granteeType, String grantorName, PrincipalType grantorType, boolean grantOption)
+            throws TException
+    {
+        GrantRevokeRoleRequest request = new GrantRevokeRoleRequest();
+        request.setRequestType(GrantRevokeType.GRANT);
+        request.setRoleName(role);
+        request.setPrincipalName(granteeName);
+        request.setPrincipalType(granteeType);
+        request.setGrantor(grantorName);
+        request.setGrantorType(grantorType);
+        request.setGrantOption(grantOption);
+        GrantRevokeRoleResponse response = client.grant_revoke_role(request);
+        if (!response.isSetSuccess()) {
+            throw new MetaException("GrantRevokeResponse missing success field");
```

Comment:
The question is if we need some extra cleanup if `client.grant_revoke_role` failed. Presumably not?